timelock-server/src/main/java/com/palantir/atlasdb/timelock/clock/ClockSkewMonitor.java

```diff
@@ -0,0 +1,176 @@
+                    } catch (Throwable t) {
+                        log.warn("ClockSkewMonitor threw an exception", t);
+                    }
+                }, SLEEP_TIME_SECONDS, TimeUnit.SECONDS
+        );
+    }
+
+    private void runInternal() {
+        monitorsByServer.forEach((server, monitor) -> {
+            long localTimeAtStart = System.nanoTime();
+            long remoteSystemTime = monitor.getSystemTimeInNanos();
+            long localTimeAtEnd = System.nanoTime();
+
+            RequestTime previousRequest = previousRequestsByServer.get(server);
+            RequestTime newRequest = new RequestTime(localTimeAtStart, localTimeAtEnd, remoteSystemTime);
+            compareClockSkew(server, newRequest, previousRequest);
+        });
+    }
+
+    private void compareClockSkew(String server, RequestTime newRequest, RequestTime previousRequest) {
+        if (previousRequest.equals(RequestTime.EMPTY)) {
+            previousRequestsByServer.put(server, newRequest);
+            return;
+        }
+
+        long maxElapsedTime = newRequest.localTimeAtEnd - previousRequest.localTimeAtStart;
+        long minElapsedTime = newRequest.localTimeAtStart - previousRequest.localTimeAtEnd;
+        long remoteElapsedTime = newRequest.remoteSystemTime - previousRequest.remoteSystemTime;
+
+        if (minElapsedTime > MAX_TIME_SINCE_PREVIOUS_REQUEST_NANOS) {
+            log.debug("It's been a long time since we last queried the server."
+                            + " Ignoring the skew, since it's not representative.",
+                    SafeArg.of("remoteElapsedTime", remoteElapsedTime));
+            previousRequestsByServer.put(server, newRequest);
+            return;
+        }
+
+        // maxElapsedTime - minElapsedTime = time for previous request and current request to complete.
+        if (maxElapsedTime - minElapsedTime > 2 * MAX_REQUEST_TIME_NANOS) {
+            log.debug("A request took too long to complete."
+                            + " Ignoring the skew, since it's not representative.",
+                    SafeArg.of("requestsDuration", maxElapsedTime - minElapsedTime));
+            previousRequestsByServer.put(server, newRequest);
+            return;
+        }
+
+        if (remoteElapsedTime < minElapsedTime || remoteElapsedTime > maxElapsedTime) {
+            long skew;
+
+            if (remoteElapsedTime < minElapsedTime) {
+                skew = minElapsedTime - remoteElapsedTime;
+            } else {
+                skew = maxElapsedTime - remoteElapsedTime;
+            }
+
+            if (skew > WARN_SKEW_THRESHOLD_NANOS && skew < ERROR_SKEW_THRESHOLD_NANOS) {
+                log.warn("Skew (in nanos) greater than expected", skew);
+            } else if (skew >= ERROR_SKEW_THRESHOLD_NANOS) {
+                log.error("Skew (in nanos) much greater than expected", skew);
+            }
+
+            METRIC_REGISTRY.histogram(String.format("clock-skew-%s-histogram", server))
```

Comment:
Might be most useful to have this be in milliseconds (could be made easier by using `Duration` as suggested above). Also, I think this is the only metric we care about, not sure how useful the others are.